File: src/main/java/me/dkim19375/mcservercreator/controller/VersionController.java
package me.dkim19375.mcservercreator.controller;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.fxml.FXML;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.ListView;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.VBox;
import me.dkim19375.mcservercreator.MCServerCreator;
import me.dkim19375.mcservercreator.util.ColorUtils;
import me.dkim19375.mcservercreator.util.ServerVersion;

import java.util.Collections;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class VersionController {
    @FXML
    private BorderPane outerBackground;
    @FXML
    private VBox background;
    @FXML
    private ListView<String> versionList;
    @FXML
    private Button submitButton;
    @FXML
    private Label mustSelectLabel;
    @FXML
    private Button backButton;
    private boolean done = false;

    @FXML
    private void initialize() {
        MCServerCreator.getInstance().setVersionController(this);
        outerBackground.setBackground(ColorUtils.getBackground(20, 20, 20));
        background.setBackground(ColorUtils.getBackground(20, 20, 20));
        submitButton.setBackground(ColorUtils.getBackground(187, 134, 252));
        versionList.setBackground(ColorUtils.getBackground(40, 40, 40));
        backButton.setBackground(ColorUtils.getBackground(3, 218, 197));
        versionList.setScaleX(1.8);
        versionList.setScaleY(1.8);
        versionList.setScaleZ(1.8);
        mustSelectLabel.setVisible(false);
        backButton.setOnAction((event) -> {
            done = false;
            MCServerCreator.getInstance().getPrimaryStage().getScene()
                    .setRoot(MCServerCreator.getInstance().getChooseTypeRoot());
        });
    }

    public void onShow() {
        final ObservableList<String> versions = FXCollections.observableArrayList();
        switch (MCServerCreator.getInstance().getServerType()) {
            case PAPER:
            case SPIGOT:
                versions.add(ServerVersion.v1_8.getVersion());
                versions.add(ServerVersion.v1_9.getVersion());
                versions.add(ServerVersion.v1_10.getVersion());
                versions.add(ServerVersion.v1_11.getVersion());
                versions.add(ServerVersion.v1_12.getVersion());
                versions.add(ServerVersion.v1_13.getVersion());
                versions.add(ServerVersion.v1_14.getVersion());
            case TUINITY:
                versions.add(ServerVersion.v1_15.getVersion());
                versions.add(ServerVersion.v1_16.getVersion());
        }
        Collections.reverse(versions);
        versionList.setItems(versions);
        versionList.setBackground(ColorUtils.getBackground(40, 40, 40));
        submitButton.setOnAction((event) -> {
            if (done) {
                return;
            }
            done = true;
            if (versionList.getSelectionModel().getSelectedItems().size() < 1) {
                final ScheduledExecutorService service = Executors.newSingleThreadScheduledExecutor();
                mustSelectLabel.setVisible(true);
                service.schedule(() -> mustSelectLabel.setVisible(false), 3, TimeUnit.SECONDS);
                return;
            }
            MCServerCreator.getInstance().setVersion(ServerVersion.fromString(versionList.getSelectionModel().getSelectedItem()));
            MCServerCreator.getInstance().getPrimaryStage().getScene().setRoot(MCServerCreator.getInstance().getChooseDirectoryRoot());
        });
    }
}

File: src/main/java/me/dkim19375/mcservercreator/controller/VersionController.kt
package me.dkim19375.mcservercreator.controller

import javafx.collections.FXCollections
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.scene.control.*
import javafx.scene.layout.BorderPane
import javafx.scene.layout.VBox
import me.dkim19375.mcservercreator.MCServerCreator
import me.dkim19375.mcservercreator.util.ColorUtils
import me.dkim19375.mcservercreator.util.ServerType
import me.dkim19375.mcservercreator.util.ServerVersion
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class VersionController {
    @FXML
    private lateinit var outerBackground: BorderPane
    @FXML
    private lateinit var background: VBox
    @FXML
    private lateinit var versionList: ListView<String>
    @FXML
    private lateinit var submitButton: Button
    @FXML
    private lateinit var mustSelectLabel: Label
    @FXML
    private lateinit var backButton: Button
    private var done = false

    @FXML
    private fun initialize() {
        MCServerCreator.instance.versionController = this
        outerBackground.background = ColorUtils.getBackground(20, 20, 20)
        background.background = ColorUtils.getBackground(20, 20, 20)
        submitButton.background = ColorUtils.getBackground(187, 134, 252)
        versionList.background = ColorUtils.getBackground(40, 40, 40)
        backButton.background = ColorUtils.getBackground(3, 218, 197)
        versionList.scaleX = 1.8
        versionList.scaleY = 1.8
        versionList.scaleZ = 1.8
        mustSelectLabel.isVisible = false
        backButton.onAction = EventHandler {
            done = false
            MCServerCreator.instance.primaryStage.scene.root = MCServerCreator.instance.chooseTypeRoot
        }
    }

    fun onShow() {
        val versions = FXCollections.observableArrayList<String?>()
        when (MCServerCreator.instance.serverType) {
            ServerType.PAPER, ServerType.SPIGOT -> {
                versions.add(ServerVersion.v1_8.version)
                versions.add(ServerVersion.v1_9.version)
                versions.add(ServerVersion.v1_10.version)
                versions.add(ServerVersion.v1_11.version)
                versions.add(ServerVersion.v1_12.version)
                versions.add(ServerVersion.v1_13.version)
                versions.add(ServerVersion.v1_14.version)
                versions.add(ServerVersion.v1_15.version)
                versions.add(ServerVersion.v1_16.version)
            }
            ServerType.TUINITY -> {
                versions.add(ServerVersion.v1_15.version)
                versions.add(ServerVersion.v1_16.version)
            }
        }
        versions.reverse()
        versionList.items = versions
        versionList.background = ColorUtils.getBackground(40, 40, 40)
        submitButton.onAction = EventHandler setOnAction@{
            if (done) {
                return@setOnAction
            }
            done = true
            if (versionList.selectionModel.selectedItems.size < 1) {
                val service = Executors.newSingleThreadScheduledExecutor()
                mustSelectLabel.isVisible = true
                service.schedule({ mustSelectLabel.isVisible = false }, 3, TimeUnit.SECONDS)
                return@setOnAction
            }
            MCServerCreator.instance
                .version = ServerVersion.fromString(versionList.selectionModel.selectedItem)!!
            MCServerCreator.instance.primaryStage.scene.root = MCServerCreator.instance.chooseDirectoryRoot
        }
    }
}